curl/tests/client_test.cc

```diff
@@ -0,0 +1,405 @@
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->cookie == "cookie_0=cookie_val_0; cookie_1=cookie_val_1");
+        }
+
+        /*
+         * ca_info tests
+         */
+        SECTION("Path to CA certificate should be unspecified by default") {
+            request test_request {"http://valid.com"};
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->cacert == "");
+        }
+
+        SECTION("cURL should receive the path to the CA certificate specified in the request") {
+            request test_request {"http://valid.com"};
+            test_client.set_ca_cert("cacert");
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->cacert == "cacert");
+        }
+
+        /*
+         * client_cert and client_key tests
+         */
+        SECTION("Client cert name should be unspecified by default") {
+            request test_request {"http://valid.com"};
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->client_cert == "");
+        }
+
+        SECTION("cURL should receive the client cert name specified in the request") {
+            request test_request {"http://valid.com"};
+            test_client.set_client_cert("cert", "key");
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->client_cert == "cert");
+        }
+
+        SECTION("Private keyfile name should be unspecified by default") {
+            request test_request {"http://valid.com"};
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->client_key == "");
+        }
+
+        SECTION("cURL should receive the private keyfile name specified in the request") {
+            request test_request {"http://valid.com"};
+            test_client.set_client_cert("cert", "key");
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->client_key == "key");
+        }
+    }
+
+    /* Exception handling tests below */
+
+    /*
+    TEST_CASE("curl::client fails to globally initialize libcurl") {
```

Comment:
The global_init_error is tricky, as global init is done as part of static initialization in the curl helper. There's currently no way to force it to be re-initialized.

For now I'd replace this block with a comment about why we don't have a test covering it.